### PR TITLE
Convert Threads ParallelReduce

### PR DESCRIPTION
**Pull request by masterleinad**
Based on top of https://github.com/kokkos/kokkos/pull/5874. Part of #5908.

### CODE REVIEW

**Review by dalg24 — APPROVED** [review]

File: core/src/Threads/Kokkos_Threads_Parallel_Range.hpp [hunk 1]
```diff
@@ -249,8 +234,7 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
         const pointer_type data =
             (pointer_type)ThreadsExec::root_reduce_scratch();
 
-        const unsigned n = Analysis::value_count(
-            ReducerConditional::select(m_functor, m_reducer));
+        const unsigned n = m_functor_reducer.get_reducer().value_count();
```
Comment (tcclevenger) [inline core/src/Threads/Kokkos_Threads_Parallel_Range.hpp:237]:
```suggestion
        const unsigned n = reducer.value_count();
```

**Review by tcclevenger — COMMENTED** [review]
A bit nitpick, feel free to ignore or incorporate into one of the PRs for the other backends.

File: core/src/Threads/Kokkos_Threads_Parallel_Range.hpp (hunk 1 above)
Comment (tcclevenger) [reply core/src/Threads/Kokkos_Threads_Parallel_Range.hpp:237]: `const ReducerType &reducer = m_functor_reducer.get_reducer();` is defined in this scope.

**Comment by masterleinad** [comment]
I decided to fix the check for the reduce result to be host-accessible for `Serial`, `Threads`, and `HPX` here as well. If preferred I can split this up again.

**Review by Rombur — APPROVED** [review]

File: core/src/Threads/Kokkos_Threads_Parallel_Range.hpp [hunk 2]
```diff
@@ -259,33 +243,19 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
...
-        m_reducer(reducer),
-        m_result_ptr(reducer.view().data()) {
-    /*static_assert( std::is_same< typename ViewType::memory_space
-                                    , Kokkos::HostSpace >::value
-      , "Reduction result on Kokkos::OpenMP must be a Kokkos::View in HostSpace"
-      );*/
+        Kokkos::Impl::MemorySpaceAccess<typename ViewType::memory_space,
+                                        Kokkos::HostSpace>::accessible,
+        "Kokkos::Threads reduce result must be a View accessible from "
+        "HostSpace");
```
Comment (dalg24) [inline core/src/Threads/Kokkos_Threads_Parallel_Range.hpp:258]: Does not compile
```
/Users/runner/work/kokkos/kokkos/core/src/Threads/Kokkos_Threads_Parallel_Range.hpp:255:50: error: use of undeclared identifier 'ViewType'
[15](https://github.com/kokkos/kokkos/actions/runs/4315922030/jobs/7530984722#step:4:16)
        Kokkos::Impl::MemorySpaceAccess<typename ViewType::memory_space,
[16](https://github.com/kokkos/kokkos/actions/runs/4315922030/jobs/7530984722#step:4:17)
                                                 ^
```

**Review by dalg24 — COMMENTED** [review]
Also not a big fan of the drive-by changes in Serial and HPX in the same PR
They are not reflected in the title line nor the description

**Comment by masterleinad** [comment]
> Also not a big fan of the drive-by changes in Serial and HPX in the same PR

Pushed it to https://github.com/masterleinad/kokkos/tree/fix_reducer_check_serial_hpx now.

File: core/src/Threads/Kokkos_Threads_Parallel_MDRange.hpp [hunk 3]
```diff
@@ -217,41 +202,28 @@ class ParallelReduce<FunctorType, Kokkos::MDRangePolicy<Traits...>, ReducerType,
...
-        m_reducer(reducer),
-        m_result_ptr(reducer.view().data()) {
-    /*static_assert( std::is_same< typename ViewType::memory_space
-                                    , Kokkos::HostSpace >::value
-      , "Reduction result on Kokkos::OpenMP must be a Kokkos::View in HostSpace"
-      );*/
+        Kokkos::Impl::MemorySpaceAccess<typename ViewType::memory_space,
+                                        Kokkos::HostSpace>::accessible,
+        "Kokkos::Threads reduce result must be a View accessible from "
+        "HostSpace");
```
Comment (dalg24) [inline core/src/Threads/Kokkos_Threads_Parallel_MDRange.hpp:226]:
```
/Users/runner/work/kokkos/kokkos/core/src/Threads/Kokkos_Threads_Parallel_MDRange.hpp:223:50: error: use of undeclared identifier 'ViewType'
[14](https://github.com/kokkos/kokkos/actions/runs/4318933367/jobs/7537807280#step:4:15)
        Kokkos::Impl::MemorySpaceAccess<typename ViewType::memory_space,

File: core/src/Threads/Kokkos_Threads_Parallel_Range.hpp [hunk 4]
```diff
@@ -249,43 +234,28 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
...
   }
 
-  template <class HostViewType>
-  ParallelReduce(const FunctorType &arg_functor, const Policy &arg_policy,
-                 const HostViewType &arg_result_view,
-                 std::enable_if_t<Kokkos::is_view<HostViewType>::value &&
-                                      !Kokkos::is_reducer<ReducerType>::value,
-                                  void *> = nullptr)
-      : m_functor(arg_functor),
+  template <class ViewType>
```
Comment (dalg24) [inline core/src/Threads/Kokkos_Threads_Parallel_Range.hpp:245]: The renaming of the template parameter was not a good change.  Not blocking but I don't like it.

Comment (masterleinad) [reply core/src/Threads/Kokkos_Threads_Parallel_Range.hpp:245]: The point of the improved `static_assert` was that the View doesn't need to live in `HostSpace` but just in a space that is host-accessible.

Comment (dalg24) [reply core/src/Threads/Kokkos_Threads_Parallel_Range.hpp:245]: And the "host" prefix convey the idea.

Comment (crtrott) [reply core/src/Threads/Kokkos_Threads_Parallel_Range.hpp:245]: Actually I am kinda ok with the change. The HostViewType could be read as it needs to be a litteral HostSpace thing, and wouldn't allow for SharedSpace for example?

**Review by crtrott — APPROVED** [review]

**Pull request merged**